nmigen_boards/mercury.py

```diff
@@ -0,0 +1,226 @@
+        # (Directly attached to FPGA)- LVCMOS33.
+        Connector("clkio", 0, "P40 P44"),  # Clock IO (Can be used as GPIO)-
+        # LVCMOS33.
+        Connector("input", 0, "P68 P97 P7 P82"),  # Input-only pins- LVCMOS33.
+        Connector("led", 0, "P13 P15 P16 P19"),  # LEDs can be used as pins
+        # as well- LVTTL.
+        Connector("pmod", 0, "P5 P4 P6 P98 P94 P93 P90 P89")  # Baseboard PMOD.
+        # Overlaps w/ GPIO bus.
+    ]
+
+    # Some default useful extensions. Attach to platform using:
+    # p.add_resources(p.leds)
+    # pmod_btn = plat.request("user_led")
+    leds = [
+        Resource("user_led", 0, Pins("1", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 1, Pins("2", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 2, Pins("3", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 3, Pins("4", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+    ]
+
+    # SRAM and 5V-tolerant I/O share a parallel bus on 200k gate version.
+    # The SRAM controller needs to take care of switching the bus between
+    # the two. Meant to be Cat() into one GPIO bus, and combined with gpio_ctl.
+    gpio_sram = [
+        Resource("gpio_sram_bus", 0,
+            Subsignal("a", Pins(""" 1  2  3  4  5  6  7  8  9 10
+                                   11 12 13 14 15 16 17 18 19 20""", dir="o",
+                                conn=("gpio", 0))),
+            # A19 is actually unused- free for GPIO
+            # 8-bit data bus
+            Subsignal("d", Pins("21 22 23 24 25 26 27 28", dir="io",
+                                conn=("gpio", 0))),
+            Subsignal("we", PinsN("29", dir="o", conn=("gpio", 0))),
+            # "Unused" is only used by GPIO, but is included here for
+            # consistency.
+            Subsignal("unused", Pins("30", dir="io", conn=("gpio", 0))),
+            # Subsignal("oe_n", Pins()),  # If OE wasn't tied to ground on
+            # Mercury, this pin would be here.
+            Attrs(IOSTANDARD="LVTTL", SLEW="FAST")
+        )
+    ]
+
+    # The "serial port" is in fact over SPI. The creators of the board provide
+    # a VHDL file for talking over this interface. In light of space
+    # constraints and the fact that both the FT245RL and FPGA can BOTH be
+    # SPI primaries, however, it may be necessary to sacrifice two "high-speed"
+    # (DIO, INPUT) pins instead.
+    serial = [
+        Resource("uart", 0,
+            Subsignal("tx", Pins("1", dir="o", conn=("dio", 0))),  # FTDI D1
+            Subsignal("rx", Pins("1", dir="i", conn=("input", 1))), # FTDI D0
+            Attrs(IOSTANDARD="LVCMOS33")
+        )
+    ]
+
+    # The remaining peripherals only make sense w/ the Baseboard installed.
+    # See: http://www.micro-nova.com/mercury-baseboard/
```

Comment:
Shouldn't this be one large group of resources called `baseboard`?

Reply:
No, because of the way SRAM works on this board, there are times you want to selectively enable certain baseboard peripherals without enabling others. In particular, only VGA and PS/2 can be enabled while still giving you access to the SRAM.

Reply:
So maybe more like `baseboard_no_sram` and `baseboard_sram` then?

Reply:
That's a good compromise<sup>1</sup>, I'll look at the schematic and see which components fit under each. To avoid repetition, I think I should keep the existing `Resource` arrays around, but maybe prefix with `_` as "internal"?

1. Btw, just as an anecdote- even with the baseboard attachment, all of Mercury's GPIO is _still_ broken out, so one can attach their own peripherals. In practice, I've found this to not work at all due to loading issues :P.

Reply:
> To avoid repetition, I think I should keep the existing `Resource` arrays around, but maybe prefix with `_` as "internal"?

Yes, that's what I would do.